synapse/federation/federation_server.py

```diff
@@ -200,6 +200,37 @@ async def on_backfill_request(
 
         return 200, res
 
+    async def on_timestamp_to_event_request(
+        self, origin: str, room_id: str, timestamp: int, direction: str
+    ) -> Tuple[int, Dict[str, Any]]:
+        """When we receive a federated `/timestamp_to_event` request,
+        handle all of the logic for validating and fetching the event.
+
+        Args:
+            origin: The server we received the event from
+            room_id: Room to fetch the event from
+            timestamp: The point in time we should navigate from in the given
+                direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            Tuple indicating the response status code and dictionary response
+            body including `event_id`.
+        """
+        with (await self._server_linearizer.queue((origin, room_id))):
+            origin_host, _ = parse_server_name(origin)
+            await self.check_server_matches_acl(origin_host, room_id)
+
+            # We only try to fetch data from the local database
+            event_id = await self.store.get_event_for_timestamp(
+                room_id, timestamp, direction
+            )
```

Comment:
Depending on the history visibility of the room, can we end up revealing the IDs (and timestamps) of events to homeservers that would otherwise be unable to view them?

Reply:
Is returning an `event_id` for an event that's hidden by `history_visibility` considered leaking?
                
`/backfill` will already return redacted versions of events that aren't visible to them in the history. I've re-used the same `filter_events_for_server` function for the same behavior.

Reply:
I wasn't aware of the precedent from `/backfill`, thank you!

If that endpoint allows a homeserver to retrieve all past `event_id`s and `origin_server_ts`s, then we don't need to use `filter_events_for_server` here.

Reply:
👍 Removed `filter_events_for_server`. We already have `check_server_matches_acl` to see if a server is allowed in the room.